components/hab/src/command/pkg/download.rs

```diff
@@ -0,0 +1,439 @@
+                RETRIES, ident, target, err
+            ))));
+        }
+
+        // At this point the artifact is in the download directory...
+        let mut artifact = PackageArchive::new(self.downloaded_artifact_path(ident, target));
+        self.verify_artifact(ui, ident, target, &mut artifact)?;
+        Ok(artifact)
+    }
+
+    // This function and it's sibling in install.rs deserve to be refactored to eke out commonality.
+    /// Retrieve the identified package from the depot, ensuring that
+    /// the artifact is downloaded.
+    fn fetch_artifact<T>(&self,
+                         ui: &mut T,
+                         ident: &PackageIdent,
+                         target: PackageTarget)
+                         -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading, format!("{} for {}", ident, target))?;
+        match self.api_client.fetch_package((ident, target),
+                                            self.token,
+                                            &self.path_for_artifact(),
+                                            ui.progress())
+        {
+            Ok(_) => Ok(()),
+            Err(api_client::Error::APIError(StatusCode::NOT_IMPLEMENTED, _)) => {
+                println!("Host platform or architecture not supported by the targeted depot; \
+                          skipping.");
+                Ok(())
+            }
+            Err(e) => Err(Error::from(e)),
+        }
+    }
+
+    fn fetch_origin_key<T>(&self,
+                           ui: &mut T,
+                           name_with_rev: &str,
+                           token: Option<&str>)
+                           -> Result<()>
+        where T: UIWriter
+    {
+        ui.status(Status::Downloading,
+                  format!("{} public origin key", &name_with_rev))?;
+        let (name, rev) = parse_name_with_rev(&name_with_rev)?;
+        self.api_client.fetch_origin_key(&name,
+                                          &rev,
+                                          token,
+                                          &self.path_for_keys(),
+                                          ui.progress())?;
+        ui.status(Status::Cached,
+                  format!("{} public origin key", &name_with_rev))?;
+        Ok(())
+    }
+
+    fn verify_artifact<T>(&self,
+                          ui: &mut T,
+                          ident: &PackageIdent,
+                          target: PackageTarget,
+                          artifact: &mut PackageArchive)
+                          -> Result<()>
+        where T: UIWriter
+    {
+        // We need to look at the artifact to know the signing keys to fetch
+        // Once we have them, it's the natural time to verify.
+        // Otherwise, it might make sense to take this fetch out of the verification code.
+        let nwr = artifact::artifact_signer(&artifact.path)?;
```

Comment:
OPTIONAL: `nwr` doesn't mean anything to me at first glance, so I think it might improve clarity to consider a different name.